daml-lf/spec/contract-id.rst

```diff
@@ -0,0 +1,245 @@
+.. Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+.. SPDX-License-Identifier: Apache-2.0
+
+V1 Contract ID allocation scheme
+================================
+
+Goals
+^^^^^
+
+* Allows ordering contract IDs and make DAML semantics depend on this
+  order, e.g., for comparison builtin and maps using IDs as keys.
+* Eliminate all contract ID translations for central commiter ledger
+
+Requirements
+^^^^^^^^^^^^
+
+**Executability**: DAML engine can determine ordering of the contract
+IDs before it hands out the transaction to the write service.
+
+**Validation**: The allocation scheme commutes with transaction
+projection. That is, the contract IDs for contracts created in the
+projection of a transaction to a set of parties can be computed
+solely from the projection and input seeds.
+
+**Unlinkability**: It is computationally infeasible to link the contract
+contents to the contract ID unless the create node is witnessed. The
+contract contents include the contract instance, the template ID, the
+stakeholders / signatories / maintainers, and the contract key.
+
+**Freshness**: It is computationally infeasible to find contracts C1 and
+C2 with the same contract ID such that one of the following holds:
+
+* The contracts have different stakeholders.
+* The contracts are created by different transactions.
+* The contracts are created by different subactions of the same
+  transaction forest where the seeds are pairwise different.
+
+**Distinctness**: The ledger can enforce that the seeds are pairwise
+different.
+
+Contract Identifiers
+^^^^^^^^^^^^^^^^^^^^
+
+A *contract identifier* (or contract ID for short) is a sequence of
+bytes defined as follows ::
+
+  ContractID := verisionPrefix ∥ discriminator ∥ suffix  
+
+where
+
+* ``∥`` is the concatenation operation; 
+* ``verisionPrefix`` is 1 byte (equal to 0) used to versioned the
+  contract ID scheme;
+* ``discriminator`` is a sequence of 32 bytes. It is like a random
+  UUID, but generated from an initial seed (called *submission seed*)
```

Comment:
```suggestion
* ``discriminator`` is a sequence of 32 bytes. It is generated from an initial seed (called *submission seed*)
```
I'd try to avoid the technical term "UUID" because there's a standard that describes the precise format. In particular, a UUID consists of exactly 16 bytes.